UNIXFSv1.md

```diff
@@ -0,0 +1,487 @@
+
+Path is a collection of path component (some bytes) seperated by `/` 0x2f, red from left to right, it is inspired by POSIX paths.
+
+Components MUST NOT contain `/` characters because else it would break the path into two components.
+
+Components SHOULD be UTF8 unicode.
+
+### Escaping
+
+The `\` may be supposed to trigger an escape sequence.
+
+This might be a thing, but is broken and inconsistent current implementations.
+So until we agree on a new spec for this, you SHOULD NOT use any escape sequence and non ascii character.
+
+
+### SHOULD NOT names
+
+Thoses names SHOULD NOT<!--MUST NOT ? in future revisions--> be used:
+
+- `.` (as this represent the self node in POSIX pathing)
+- `..` (as this represent the parrent node in POSIX pathing)
+- nothing (<!--TODO: check that this is true-->we don't actually know the failure mode for this, but it really feels like this shouldn't be a thing)
+
+## How to read a Node
+
+First you get some CID, this will be what will we be trying to decode.
+
+For recap, every CIDs MUST include:
+1. A [multicodec](https://github.com/multiformats/multicodec), also called codec.
+1. A [Multihash](https://github.com/multiformats/multihash) (used to specify a hashing algorithm, some hashing parameters and some digest)
+
+### Get the block
+
+The first step is to get the block, that means the actual bytes which hashed by the multihash inside the CID give you the same multihash back.
+
+### Decoding the bytes
+
+#### Block limit
+
+A so called "block limit" is in place, we do not allow any single block to be bigger than 2MiB.
+
+Implementation SHOULD try to not emit 1MiB bigger blocks, but MUST decode blocks <= 2MiB.
+
+#### Multicodecs
+
+With Unixfs we deal with two codecs which will be decoded differently:
+- `Raw`, single blocks files
+- `Dag-PB`, any nodes
+
+##### `Raw` blocks
+
+The most simplest node is a `Raw` node.
+
+They are always of type file.
+
+They can be recognised because their CIDs have `Raw` codec.
+
+The file content is purely the block body.
+
+They never have any childs, and thus are also known as single block files.
+
+Their size (both `dagsize` and `blocksize`) is the length of the block body.
+
+###### `Raw` Example
+
+Let's build a `Raw` file whoses content is `test`.
+
+1. First hash the data:
+```console
+$ echo -n "test" | sha256sum
+9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08  -
+```
+
+2. Add the CID header:
+```
+f this is the multibase prefix, we need it because we are working with a hex CID, this is omitted for binary CIDs
+ 01 the CID version, here one
+   55 the codec, here we MUST use Raw because this is a Raw file
+     12 the hashing function used, here sha256
+       20 the digest length 32 bytes
+         9f86d081884c7d659a2feaa0c55ad015a3bf4f1b2b0b822cd15d6c15b0f00a08 the digest we computed earlier
+```
+
+3. Profit
+Assuming we stored this block in some implementation of our choice which makes it accessible to our client, we can try to decode it:
```

Comment:
some implementation of our choice of what?
you might want to refer to `kubo` directly, as an IPFS implementation with a datastore that persists data locally, and implements UnixFS.

Reply:
I didn't wanted to include poking into flatfs or creating a car file as part of the example.

I can add `echo -n "test" | ipfs block put` if you want.